*** id2.c	Tue Apr 14 10:59:39 2015
--- id2.c.new	Tue Apr 14 10:59:30 2015
*************** ID2ReplyAsnWrite(ID2ReplyPtr ptr, AsnIoP
*** 2972,2978 ****
           goto erret;
        }
     }
!    if (ptr -> discard != -1) {
        av.intvalue = ptr -> discard;
        retval = AsnWrite(aip, ID2_REPLY_discard,  &av);
     }
--- 2972,2978 ----
           goto erret;
        }
     }
!    if (ptr -> discard != (Uint2)-1) {
        av.intvalue = ptr -> discard;
        retval = AsnWrite(aip, ID2_REPLY_discard,  &av);
     }
*************** ID2ReplyGetBlobIdAsnWrite(ID2ReplyGetBlo
*** 3793,3800 ****
        av.boolvalue = ptr -> end_of_reply;
        retval = AsnWrite(aip, REPLY_GET_BLOB_ID_end_of_reply,  &av);
     }
!    av.intvalue = ptr -> blob_state;
!    retval = AsnWrite(aip, REPLY_GET_BLOB_ID_blob_state,  &av);
     if (! AsnCloseStruct(aip, atp, (Pointer)ptr)) {
        goto erret;
     }
--- 3793,3802 ----
        av.boolvalue = ptr -> end_of_reply;
        retval = AsnWrite(aip, REPLY_GET_BLOB_ID_end_of_reply,  &av);
     }
!    if (ptr -> blob_state != 0) {
!        av.intvalue = ptr -> blob_state;
!        retval = AsnWrite(aip, REPLY_GET_BLOB_ID_blob_state,  &av);
!    }
     if (! AsnCloseStruct(aip, atp, (Pointer)ptr)) {
        goto erret;
     }
*************** ID2ReplyGetBlobAsnWrite(ID2ReplyGetBlobP
*** 4150,4157 ****
           goto erret;
        }
     }
!    av.intvalue = ptr -> blob_state;
!    retval = AsnWrite(aip, ID2_REPLY_GET_BLOB_blob_state,  &av);
     if (! AsnCloseStruct(aip, atp, (Pointer)ptr)) {
        goto erret;
     }
--- 4152,4161 ----
           goto erret;
        }
     }
!    if (ptr -> blob_state != 0) {
!        av.intvalue = ptr -> blob_state;
!        retval = AsnWrite(aip, ID2_REPLY_GET_BLOB_blob_state,  &av);
!    }
     if (! AsnCloseStruct(aip, atp, (Pointer)ptr)) {
        goto erret;
     }
*************** ID2SReplyGetSplitInfoAsnWrite(ID2SReplyG
*** 4524,4531 ****
           goto erret;
        }
     }
!    av.intvalue = ptr -> blob_state;
!    retval = AsnWrite(aip, GET_SPLIT_INFO_blob_state,  &av);
     if (! AsnCloseStruct(aip, atp, (Pointer)ptr)) {
        goto erret;
     }
--- 4528,4537 ----
           goto erret;
        }
     }
!    if (ptr -> blob_state != 0) {
!        av.intvalue = ptr -> blob_state;
!        retval = AsnWrite(aip, GET_SPLIT_INFO_blob_state,  &av);
!    }
     if (! AsnCloseStruct(aip, atp, (Pointer)ptr)) {
        goto erret;
     }
